platform/bukkit/src/main/java/me/lokka30/treasury/plugin/bukkit/hooks/miniplaceholders/MiniPlaceholdersHookRegistrar.java

```diff
@@ -0,0 +1,39 @@
+/*
+ * This file is/was part of Treasury. To read more information about Treasury such as its licensing, see <https://github.com/ArcanePlugins/Treasury>.
+ */
+
+package me.lokka30.treasury.plugin.bukkit.hooks.miniplaceholders;
+
+import io.github.miniplaceholders.api.Expansion;
+import me.lokka30.treasury.plugin.bukkit.TreasuryBukkit;
+import me.lokka30.treasury.plugin.bukkit.hooks.Hook;
+import me.lokka30.treasury.plugin.bukkit.vendor.BukkitVendor;
+import org.jetbrains.annotations.NotNull;
+
+public class MiniPlaceholdersHookRegistrar implements Hook {
+
+    private Expansion expansion;
+
+    @Override
+    public @NotNull String getPlugin() {
+        return "MiniPlaceholders";
+    }
+
+    @Override
+    public boolean register(@NotNull final TreasuryBukkit plugin) {
+        if (!BukkitVendor.isPaper()) { // Paper, Folia
+            return false;
+        }
+        // TODO: Seek help from 4drian3d !!!
```

Comment:
@4drian3d how can we register the placeholders just like in PAPI? 
plz help k thx luv <3

Reply:
> @4drian3d how can we register the placeholders just like in PAPI? plz help k thx luv <3
```java
Expansion.builder("expansionname")
  // This is a filter that ensures that any audiences received by Audience or Relational placeholders are of the specified class, this is optional, and global placeholders are not affected
  .filter(Player.class)
  .audiencePlaceholder("placeholdername", (audience, argumentQueue, context) -> {
    // This is a placeholder that requires a player to be present to be executed
    // Placeholder: <expansionname_placeholdername>
    final Player player = (Player) audience;
    return Tag.selfClosingInserting(player.getDisplayname());
  })
  .globalPlaceholder("anotherplaceholder", (argumentQueue, context) -> {
    // This will be a placeholder that does not require any player to be executed
    // Placeholder: <expansionname_anotherplaceholder>
    return Tag.styling(NameTextColor.RED);
  })
  .relationalPlaceholder("relationalplaceholder", (audience, relationalAudience, argumentQueue, context) -> 
    // This is a relational placeholder that returns a value based on a relationship between 2 players. The concept is similar to PlaceholderAPI relational placeholders
    final Player player = (Player) player;
    final Player relationalPlayer = (Player) relationalAudience;
    return Tag.selfClosingInserting(API.areEnemy(player, relationalPlayer) ? text("Enemy", NamedTextColor.RED) : text("Ally", NamedTextColor.AQUA));
  })
  // With this method you build the expansion, from here you can decide if the expansion will be used locally in your plugin or if you want to register it
  .build()
  // From here on, the expansion is registered globally and can be used from any plugin that is compatible with MiniPlaceholders
  .register();
```

https://github.com/MiniPlaceholders/MiniPlaceholders/wiki/Expansion-Creation
https://github.com/MiniPlaceholders/MiniPlaceholders/wiki/Integration

Reply:
Sure, but we have a ton of eco placeholders, declaring them 1 by 1 will get real messy real quick.

Reply:
> Sure, but we have a ton of eco placeholders, declaring them 1 by 1 will get real messy real quick.

This way of registering placeholders in MiniPlaceholders allows for better code readability and to know instantly in which part a placeholder is registered, along with having a different API of arguments that is not based on plain text, but based on the MiniMessage API that allows an easier creation of placeholders with arguments.
If you want to reduce the amount of code used only in the creation of placeholders, you can make separate placeholder classes as in the [Spark expansion](https://github.com/MiniPlaceholders/Spark-Expansion/tree/main/common/src/main/java/io/github/miniplaceholders/expansion/spark/common)

Reply:
What if I do something like what you did in the [PAPI expansion](https://github.com/MiniPlaceholders/PlaceholderAPI-Expansion/blob/main/paper/src/main/java/io/github/miniplaceholders/expansion/placeholderapi/paper/PaperPlugin.java#L26) ?

Reply:
> What if I do something like what you did in the [PAPI expansion](https://github.com/MiniPlaceholders/PlaceholderAPI-Expansion/blob/main/paper/src/main/java/io/github/miniplaceholders/expansion/placeholderapi/paper/PaperPlugin.java#L26) ?

If you really want to do something similar, you could make a placeholder `<treasury_eco:argument>` and in the argument put all the required data, that way it would look like a PlaceholderAPI placeholder, but I wouldn't recommend it

```java
Expansion.builder("treasury")
  .audienceExpansion("eco", (audience, argumentQueue, context) -> {
    final Argument argument = argumentQueue.popOr("You need to provide an argument");
    return processEcoPlaceholder(audience, argument);
  })
  .build()
  .register();

public Tag processEcoPlaceholder(Audience audience, Argument argument) {
  final Player player = (Player) audience;
  final String argumentString = argument.value();
  // some process
  return Tag.preProcessParsed("Some value");
}
```

Reply:
Well it's gonna look like that since 90% of treasury's placeholders require additional input. For example, taking specific player's balance looks like `%treasury_eco_balance_MrIvanPlays%`, in MiniPlaceholders it'd look like `<treasury_eco:balance_MrIvanPlays>`. It's something I'm willing to sacrifice to save time and useless effort. But in any case, if I put that time it'd still have to look something like `<treasury_eco_balance:MrIvanPlays>`.